src/execution.jl

```diff
@@ -356,92 +423,51 @@ The output of this function is automatically cached, i.e. you can simply call `c
-                jit_options[CUDAdrv.JIT_GENERATE_DEBUG_INFO] = true
-            end
-
-            # link the CUDA device library
-            image = asm
-            # linking the device runtime library requires use of the CUDA linker,
-            # which in turn switches compilation to device relocatable code (-rdc) mode.
-            #
-            # even if not doing any actual calls that need -rdc (i.e., calls to the runtime
-            # library), this significantly hurts performance, so don't do it unconditionally
-            intrinsic_fns = ["vprintf", "malloc", "free", "__assertfail",
-                            "__nvvm_reflect" #= TODO: should have been optimized away =#]
-            if !isempty(setdiff(undefined_fns, intrinsic_fns))
-                @timeit_debug to "device runtime library" begin
-                    linker = CUDAdrv.CuLink(jit_options)
-                    CUDAdrv.add_file!(linker, libcudadevrt[], CUDAdrv.JIT_INPUT_LIBRARY)
-                    CUDAdrv.add_data!(linker, kernel_fn, asm)
-                    image = CUDAdrv.complete(linker)
-                end
-            end
-
-            # JIT into an executable kernel object
-            mod = CuModule(image, jit_options)
-            fun = CuFunction(mod, kernel_fn)
-            kernel = HostKernel{f,tt}(ctx, mod, fun)
-
-            create_exceptions!(mod)
-
-            compilecache[key] = kernel
-            stop = time_ns()
-            @debug begin
-                ver = version(kernel)
-                mem = memory(kernel)
-                reg = registers(kernel)
-                fn = something(name, nameof(f))
-                """Compiled $fn($(join(tt.parameters, ", "))) to PTX $(ver.ptx) for SM $(ver.binary) in $(round((time_ns() - start) / 1000000; digits=2)) ms.
-                   Kernel uses $reg registers, and $(Base.format_bytes(mem.local)) local, $(Base.format_bytes(mem.shared)) shared, and $(Base.format_bytes(mem.constant)) constant memory."""
-            end
-        end
-
-        return compilecache[key]::HostKernel{f,tt}
-    end
+@generated function cufunction(f::Core.Function, tt::Type=Tuple{}; kwargs...)
+    # generated function that crafts a custom code info to call the actual cufunction impl.
+    # this gives us the flexibility to insert manual back edges for automatic recompilation.
+    tt = tt.parameters[1]
+
+    # get a hold of the method and code info of the kernel function
+    sig = Tuple{f, tt.parameters...}
+    mthds = _methods_by_ftype(sig, -1, typemax(UInt))
+    length(mthds) == 1 || return (:(throw(MethodError(f,tt))))
+    mtypes, msp, m = mthds[1]
+    mi = ccall(:jl_specializations_get_linfo, Ref{MethodInstance}, (Any, Any, Any), m, mtypes, msp)
+    mi.def.isva && return :(error("varargs kernel methods are not supported"))
+    ci = retrieve_code_info(mi)
+    @assert isa(ci, CodeInfo)
+
+    # generate a unique id to represent this specialization
+    global specialization_counter
```

Comment:
Hm, shouldn't this be the world-age of the `mi` instead of the many invocations of the generator?

Reply:
But the method age of the `mi` doesn't increase when a called method gets redefined, which was the reason for this PR in the first place.

Reply:
Hm I thought we updated the upper bound? Maybe I am simply not understanding the caching scheme you employed here.
A `@generated` function may be invoked by the compiler as often as it wants, which the condition that the same inputs yield the same output (within a worldarg range).

Instead of looking at `mi.min_world` have you looked at `mi.max_world`? That should get updated, when there is a method instance with a signature in a newer world.

Reply:
The parent world counters don't change when redefining child functions.

```
julia> function method_age(f, tt)
           sig = Base.signature_type(f, tt)
           mthds = _methods_by_ftype(sig, -1, typemax(UInt))
           Base.isdispatchtuple(tt) || return(:(error("$tt is not a dispatch tuple")))
           length(mthds) == 1 || return (:(throw(MethodError(f,tt))))
           mtypes, msp, m = mthds[1]
           mi = ccall(:jl_specializations_get_linfo, Ref{MethodInstance}, (Any, Any, Any), m, mtypes, msp)
           mi.def.isva && return :(error("varargs kernel methods are not supported"))
           ci = retrieve_code_info(mi)
           @assert isa(ci, CodeInfo)

           dump(stderr, ci)
       end
method_age (generic function with 1 method)

julia> foo() = bar()
foo (generic function with 1 method)

julia> bar() = 1
bar (generic function with 1 method)

julia> method_age(foo, Tuple{})
CodeInfo
  parent: MethodInstance
    def: Method
      primary_world: UInt64 0x0000000000006b26
      deleted_world: UInt64 0xffffffffffffffff
      specializations: Core.TypeMapEntry
        min_world: UInt64 0x0000000000000001
        max_world: UInt64 0xffffffffffffffff
  min_world: UInt64 0x0000000000000001
  max_world: UInt64 0xffffffffffffffff

julia> method_age(bar, Tuple{})
CodeInfo
  parent: MethodInstance
    def: Method
      primary_world: UInt64 0x0000000000006b27
      deleted_world: UInt64 0xffffffffffffffff
      specializations: Core.TypeMapEntry
        min_world: UInt64 0x0000000000000001
        max_world: UInt64 0xffffffffffffffff
  min_world: UInt64 0x0000000000000001
  max_world: UInt64 0xffffffffffffffff

julia> bar() = 2
bar (generic function with 1 method)

julia> method_age(foo, Tuple{})
CodeInfo
  parent: MethodInstance
    def: Method
      primary_world: UInt64 0x0000000000006b26
      deleted_world: UInt64 0xffffffffffffffff
      specializations: Core.TypeMapEntry
        min_world: UInt64 0x0000000000000001
        max_world: UInt64 0xffffffffffffffff
  min_world: UInt64 0x0000000000000001
  max_world: UInt64 0xffffffffffffffff

julia> method_age(bar, Tuple{})
CodeInfo
  parent: MethodInstance
    def: Method
      primary_world: UInt64 0x0000000000006b28
      deleted_world: UInt64 0xffffffffffffffff
      specializations: Core.TypeMapEntry
        min_world: UInt64 0x0000000000000001
        max_world: UInt64 0xffffffffffffffff
  min_world: UInt64 0x0000000000000001
  max_world: UInt64 0xffffffffffffffff
```